File: packages/client-sdk/examples/risk-scoring.ts
// examples/risk-scoring.ts
// AMTTP SDK Examples - Risk Scoring

import { AMTTPClient, MLService, PolicyAction, RiskLevel, utils } from '@amttp/client-sdk';

// ============================================================
// Example 1: Basic Risk Scoring
// ============================================================

async function basicScoring() {
  console.log('\n[Example 1: Basic Risk Scoring]');
  
  // Create ML service
  const mlService = new MLService('http://localhost:8000');
  
  // Score a transaction
  const risk = await mlService.scoreTransaction('tx_001', {
    amount: 1.5,          // ETH
    velocity_24h: 3,
    account_age_days: 90,
    country_risk: 0.2,
  });
  
  console.log(`Risk Score: ${utils.formatRiskScore(risk.riskScore)}`);
  console.log(`Risk Level: ${RiskLevel[risk.riskLevel]}`);
  console.log(`Action: ${PolicyAction[risk.action]}`);
  console.log(`Model: ${risk.modelVersion}`);
  console.log('Recommendations:');
  risk.recommendations.forEach(rec => console.log(`  - ${rec}`));
}

// ============================================================
// Example 2: Full Client Integration
// ============================================================

async function fullClientIntegration() {
  console.log('\n[Example 2: Full Client Integration]');
  
  const client = new AMTTPClient({
    rpcUrl: 'https://mainnet.infura.io/v3/YOUR_KEY',
    contractAddress: '0x...',
    oracleUrl: 'http://localhost:3000',
    mlApiUrl: 'http://localhost:8000',
    // privateKey: '0x...', // Optional for signing
  });
  
  // Score transaction risk
  const risk = await client.scoreTransactionRisk({
    from: '0x1234...',
    to: '0x5678...',
    amount: 10,  // ETH
    metadata: {
      velocity_24h: 5,
      account_age_days: 30,
    }
  });
  
  console.log(`Risk Category: ${risk.riskCategory}`);
  console.log(`Confidence: ${(risk.confidence * 100).toFixed(1)}%`);
  
  // Decide action based on risk
  if (risk.riskScore >= 0.7) {
    console.log('⚠️ High risk - Use escrow protection');
  } else if (risk.riskScore >= 0.4) {
    console.log('👀 Medium risk - Requires review');
  } else {
    console.log('✅ Low risk - Proceed normally');
  }
}

// ============================================================
// Example 3: Batch Scoring
// ============================================================

async function batchScoring() {
  console.log('\n[Example 3: Batch Scoring]');
  
  const mlService = new MLService('http://localhost:8000');
  
  const transactions = [
    { id: 'tx_001', features: { amount: 1.0, velocity_24h: 1 } },
    { id: 'tx_002', features: { amount: 5.0, velocity_24h: 10 } },
    { id: 'tx_003', features: { amount: 50.0, velocity_24h: 50 } },
  ];
  
  const risks = await mlService.scoreBatch(transactions);
  
  console.log('Batch Results:');
  risks.forEach((risk, i) => {
    const color = utils.getRiskColor(risk.riskCategory);
    console.log(
      `  ${transactions[i].id}: ` +
      `${utils.formatRiskScore(risk.riskScore)} -> ` +
      `${PolicyAction[risk.action]}`
    );
  });
}

// ============================================================
// Example 4: React Hook Pattern
// ============================================================

/*
// hooks/useRiskScore.ts
import { useState, useCallback } from 'react';
import { MLService, RiskScore } from '@amttp/client-sdk';

const mlService = new MLService(process.env.NEXT_PUBLIC_ML_API_URL);

export function useRiskScore() {
  const [loading, setLoading] = useState(false);
  const [risk, setRisk] = useState<RiskScore | null>(null);
  const [error, setError] = useState<Error | null>(null);

  const scoreTransaction = useCallback(async (
    to: string,
    amount: number,
    features?: Record<string, number>
  ) => {
    setLoading(true);
    setError(null);
    
    try {
      const result = await mlService.scoreTransaction(
        `tx_${Date.now()}`,
        { amount, ...features }
      );
      setRisk(result);
      return result;
    } catch (err) {
      setError(err as Error);
      throw err;
    } finally {
      setLoading(false);
    }
  }, []);

  return { scoreTransaction, risk, loading, error };
}

// Usage in component:
function TransferForm() {
  const { scoreTransaction, risk, loading } = useRiskScore();
  
  const handleSubmit = async (to: string, amount: number) => {
    const riskResult = await scoreTransaction(to, amount);
    
    if (riskResult.action === PolicyAction.BLOCK) {
      alert('Transaction blocked due to high risk');
      return;
    }
    
    if (riskResult.action === PolicyAction.ESCROW) {
      // Use escrow flow
    }
    
    // Proceed with transaction
  };
}
*/

// ============================================================
// Example 5: Express.js Middleware
// ============================================================

/*
// middleware/riskCheck.ts
import { Request, Response, NextFunction } from 'express';
import { MLService, PolicyAction } from '@amttp/client-sdk';

const mlService = new MLService(process.env.ML_API_URL);

export async function riskCheckMiddleware(
  req: Request,
  res: Response,
  next: NextFunction
) {
  const { to, amount, from } = req.body;
  
  if (!to || !amount) {
    return next();
  }
  
  try {
    const risk = await mlService.scoreTransaction(`api_${Date.now()}`, {
      amount: Number(amount),
      from_address: from,
      to_address: to,
    });
    
    // Attach risk to request
    req.riskScore = risk;
    
    // Block high-risk transactions
    if (risk.action === PolicyAction.BLOCK) {
      return res.status(403).json({
        error: 'Transaction blocked due to high risk',
        riskScore: risk.riskScore,
        recommendations: risk.recommendations,
      });
    }
    
    next();
  } catch (error) {
    // Allow transaction but log warning
    console.warn('Risk check failed:', error);
    next();
  }
}

// Usage:
// app.post('/api/transfer', riskCheckMiddleware, transferHandler);
*/

// ============================================================
// Run Examples
// ============================================================

async function main() {
  console.log('='.repeat(60));
  console.log('AMTTP TypeScript SDK Examples');
  console.log('='.repeat(60));
  
  try {
    await basicScoring();
    // await fullClientIntegration();
    // await batchScoring();
  } catch (error) {
    console.error('Example failed:', error);
  }
}

main();

File: packages/python-sdk/examples/risk_scoring.py
"""
AMTTP SDK Examples - Risk Scoring

Demonstrates ML-powered fraud detection usage.
"""
from amttp import AMTTPClient, AMTTPConfig, PolicySettings, PolicyAction

# ============================================================
# Example 1: Basic Risk Scoring
# ============================================================

def basic_scoring():
    """Score a single transaction."""
    
    # Configure client (can also use AMTTPConfig.from_env())
    config = AMTTPConfig(
        rpc_url="https://mainnet.infura.io/v3/YOUR_KEY",
        oracle_url="http://localhost:3000",
        ml_api_url="http://localhost:8000",
    )
    
    client = AMTTPClient(config)
    
    # Score a transaction
    risk = client.score_transaction(
        to="0x742d35Cc6634C0532925a3b844Bc9e7595f9EBEB",
        value=1_000_000_000_000_000_000,  # 1 ETH
        features={
            "velocity_24h": 3,
            "account_age_days": 90,
            "country_risk": 0.2,
        }
    )
    
    print(f"Risk Score: {risk.risk_score:.2%}")
    print(f"Risk Level: {risk.risk_level.name}")
    print(f"Action: {risk.action.name}")
    print(f"Confidence: {risk.confidence:.2%}")
    print(f"Model: {risk.model_version}")
    print("\nRecommendations:")
    for rec in risk.recommendations:
        print(f"  - {rec}")


# ============================================================
# Example 2: Batch Scoring
# ============================================================

def batch_scoring():
    """Score multiple transactions at once."""
    
    config = AMTTPConfig.from_env()
    client = AMTTPClient(config)
    
    transactions = [
        {"to": "0x111...", "value": 1e18, "features": {"velocity_24h": 1}},
        {"to": "0x222...", "value": 5e18, "features": {"velocity_24h": 10}},
        {"to": "0x333...", "value": 50e18, "features": {"velocity_24h": 50}},
    ]
    
    risks = client.score_batch(transactions)
    
    for i, risk in enumerate(risks):
        print(f"Transaction {i+1}: {risk.risk_score:.2%} -> {risk.action.name}")


# ============================================================
# Example 3: Policy Validation
# ============================================================

def validate_with_policy():
    """Validate transaction against on-chain policy."""
    
    config = AMTTPConfig(
        rpc_url="https://mainnet.infura.io/v3/YOUR_KEY",
        policy_manager_contract="0x...",  # Your deployed contract
        oracle_url="http://localhost:3000",
        ml_api_url="http://localhost:8000",
        private_key="0x...",  # Your private key
    )
    
    client = AMTTPClient(config)
    
    # Validate transaction
    result = client.validate_transaction(
        to="0x742d35Cc6634C0532925a3b844Bc9e7595f9EBEB",
        value=10_000_000_000_000_000_000,  # 10 ETH
    )
    
    if result.success:
        print(f"✅ Transaction allowed")
        print(f"   Action: {result.action_taken.name}")
        
        if result.action_taken == PolicyAction.ESCROW:
            print(f"   Escrow ID: {result.escrow_id}")
            print("   Funds will be held until release conditions met")
    else:
        print(f"❌ Transaction blocked: {result.error}")


# ============================================================
# Example 4: Set User Policy
# ============================================================

def set_user_policy():
    """Configure custom policy on-chain."""
    
    config = AMTTPConfig(
        rpc_url="https://mainnet.infura.io/v3/YOUR_KEY",
        policy_manager_contract="0x...",
        private_key="0x...",
    )
    
    client = AMTTPClient(config)
    
    # Define policy
    policy = PolicySettings(
        max_amount=100 * 10**18,   # 100 ETH max per transaction
        daily_limit=50 * 10**18,   # 50 ETH daily limit
        weekly_limit=200 * 10**18,
        monthly_limit=500 * 10**18,
        risk_threshold=600,        # Block transactions with risk > 0.60
        auto_approve=True,         # Auto-approve low risk
        cooldown_period=3600,      # 1 hour between large transactions
    )
    
    # Set policy on-chain
    tx_hash = client.set_policy(policy)
    print(f"Policy updated: {tx_hash}")


# ============================================================
# Example 5: Integration with Web Application
# ============================================================

def web_app_integration():
    """Example FastAPI integration."""
    
    from fastapi import FastAPI, HTTPException
    from pydantic import BaseModel
    
    app = FastAPI()
    
    # Initialize client once
    config = AMTTPConfig.from_env()
    amttp = AMTTPClient(config)
    
    class TransactionRequest(BaseModel):
        to: str
        value: int  # in wei
        velocity_24h: int = 1
        account_age_days: int = 30
    
    @app.post("/api/check-transaction")
    async def check_transaction(req: TransactionRequest):
        # Score the transaction
        risk = amttp.score_transaction(
            to=req.to,
            value=req.value,
            features={
                "velocity_24h": req.velocity_24h,
                "account_age_days": req.account_age_days,
            }
        )
        
        # Return assessment
        return {
            "allowed": risk.action != PolicyAction.BLOCK,
            "risk_score": risk.risk_score,
            "action": risk.action.name,
            "recommendations": risk.recommendations,
        }


# ============================================================
# Example 6: Health Check
# ============================================================

def health_check():
    """Check all services status."""
    
    config = AMTTPConfig.from_env()
    client = AMTTPClient(config)
    
    status = client.health_check()
    
    print("Service Status:")
    print(f"  Web3 Connected: {status['web3_connected']}")
    print(f"  Chain ID: {status['chain_id']}")
    print(f"  Address: {status['address']}")
    print(f"  ML API: {status['ml_api']['status']}")
    print(f"  Oracle: {status['oracle']['status']}")


if __name__ == "__main__":
    print("=" * 60)
    print("AMTTP SDK Examples")
    print("=" * 60)
    
    # Run basic scoring example
    print("\n[Example 1: Basic Scoring]")
    basic_scoring()
